Signing Elasticsearch Requests using AWS SDK

nshumoogum opened the pull request:
### What

Provide working solution to sign Http requests to AWS managed elasticsearch service using AWS SDK to retrieve credentials using several known providers and the V4 AWS SDK signer method.

Total removal of smartystreets sign method

awsauth package will sit here until we decide to move it to new library, solution is not Elasticsearch specific.

### How to review

- Check changes make sense
- Unit tests pass
- Check documentation in README.md is clear

### Who can review

Anyone

nshumoogum committed:
Update healthcheck method to use new awsauth package for signing http request

nshumoogum committed:
Use aws sdk v2 library to sign commits

nshumoogum committed:
Update aws signing package to use 3 possible providers

Part of signing an http request the auth package will create a chain of providers.

1) Provider will attempt to use environment variables
2) Provider will attempt to use a profile from a credentials file
3) Provider will attempt to retrieve credentials whilst on an EC2 instance

nshumoogum committed:
Tidy up new awsauth package

Also update elasticsearch client to contain aws session. The session
should be created once by an application on startup, otherwise performance
issues will result in slow requests.

nshumoogum committed:
Request to elasticsearch cluster health should ot retry if it fails.

Health checker is on a separate ticker which will retry the request to elasticsearch if it fails.

nshumoogum committed:
Extend documentation to include awsauth package

nshumoogum committed:
Fix tests and refactor duplication of code

ChrysmOre commented on awsauth/sign.go:
```diff
@@ -6,46 +6,71 @@ import (
-			return err
+func (s *Signer) Sign(req *http.Request, bodyReader io.ReadSeeker, currentTime time.Time) (err error) {
+	if s != nil && s.v4 != nil {
+		_, err = s.v4.Sign(req, bodyReader, s.awsService, s.awsRegion, time.Now())
```
Could probably be simplified to one line `if _, err = s.v4.Sign(...); err != nil{`

ChrysmOre commented on awsauth/sign.go:
```diff
@@ -6,46 +6,71 @@ import (
+				Client: ec2metadata.New(sess),
+			},
+		})
+
```
Small thing but bracket should probably closed on same column as it was opened (on next line)

nshumoogum committed:
Remove smartystreets go aw auth package from signing http requests

nshumoogum committed:
Fix unit tests to create access ID and secret key before running tests

cadmiumcat commented on README.md:
````diff
@@ -37,12 +41,93 @@ Call elasticsearch health checker with `cli.Checker(context.Background())` and t
+        ... // Handle error
+    }
+...
+```
````
Use tab to align this paragraph so that it's under 1) ➡️ 
I also like to use syntax highlighting 
(sorry if the suggestion makes no sense - sometimes it gets messed up with the many `)
```suggestion
    Requires `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables to be set/exported onto the environment.
    ```go
    import esauth "github.com/ONSdigital/dp-elasticsearch/v2/awsauth"
    ...
        signer, err := esauth.NewAwsSigner("", "", "eu-west-1", "es")
        if err != nil {
            ... // Handle error
        }
    ...
    ```
```

cadmiumcat commented on README.md:
````diff
@@ -37,12 +41,93 @@ Call elasticsearch health checker with `cli.Checker(context.Background())` and t
+        ...
+    }
+...
+```
````
➡️  and syntax highlighting
```suggestion
    Requires credentials file to exist in the location specified in NewAwsSigner func.
    File must contain the keys necessary under the matching Profile heading, see example below:

    ```
    [development]
    aws_access_key_id=<access key id>
    aws_secret_access_key=<secret access key>
    region=<region>
    ```

    ```go
    import esauth "github.com/ONSdigital/dp-elasticsearch/v2/awsauth"
    ...
        signer, err := esauth.NewAwsSigner("~/.aws/credentials", "development", "eu-west-1", "es")
        if err != nil {
            ...
        }
    ...
    ```
```

cadmiumcat commented on README.md:
````diff
@@ -37,12 +41,93 @@ Call elasticsearch health checker with `cli.Checker(context.Background())` and t
+The signer object should be created once on startup of an application and reused for each request, otherwise you will experience performance issues due to creating a session for every request.
+
+To sign elasticsearch requests, one can use the signer like so:
+```
````
syntax highlighting
```suggestion
```go
```

cadmiumcat commented on README.md:
```diff
@@ -37,12 +41,93 @@ Call elasticsearch health checker with `cli.Checker(context.Background())` and t
+### awsauth package
+
+Using AWS SDK library to create a signer function to successfully sign elasticsearch requests hosted in AWS.
+The function adds multiple providers to the credentials chain that is used by the aws sdk v4 signer method `Sign`.
```
🔠 uppercase for consistency 😁 

```suggestion
The function adds multiple providers to the credentials chain that is used by the [AWS SDK v4 signer method `Sign`](https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/#Signer.Sign).
```

cadmiumcat commented on README.md:
````diff
@@ -37,12 +41,93 @@ Call elasticsearch health checker with `cli.Checker(context.Background())` and t
+...
+```
+
+3) **EC2 Role Provider** will attempt to retrieve credentials using an EC2 metadata client (this is created using an aws sdk session).
````
🔠 
```suggestion
3) **EC2 Role Provider** will attempt to retrieve credentials using an EC2 metadata client (this is created using an AWS SDK session).
```

cadmiumcat commented on awsauth/sign.go:
```diff
@@ -6,46 +6,70 @@ import (
-	if signer.awsRegion == "" {
+func validateAwsSDKSigner(awsRegion, awsService string) error {
+	if awsRegion == "" {
 		return errors.New("No AWS region was provided. Cannot sign request.")
```
🔡 lowercase for consistency. Similarly with other `errors.New`
```suggestion
		return errors.New("no AWS region was provided. Cannot sign request.")
```

cadmiumcat reviewed (commented):
The code changes make sense to me, and so do the docs. I've added some minor formatting suggestions
`make test` and `make audit` pass

nshumoogum committed:
Add syntax highlighting to documentation and realign some of the doc

Also update error messages from awsauth package, lower case starting character and remove final fullstop. This allows the ability to better wrap error for apps and other libraries using this package.

cadmiumcat reviewed (approved):
This looks good to me 🤞
